pkg/deploy/migration/on-reconcile-one-time-migration.go

```diff
@@ -0,0 +1,285 @@
+// available under the terms of the Eclipse Public License 2.0
+// which is available at https://www.eclipse.org/legal/epl-2.0/
+//
+// SPDX-License-Identifier: EPL-2.0
+//
+// Contributors:
+//   Red Hat, Inc. - initial API and implementation
+//
+
+package migration
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/eclipse-che/che-operator/pkg/deploy"
+	"github.com/eclipse-che/che-operator/pkg/util"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type Migrator struct {
+	deploy.Reconcilable
+
+	migrationDone bool
+}
+
+func NewMigrator() *Migrator {
+	return &Migrator{
+		migrationDone: false,
+	}
+}
+
+func (m *Migrator) Reconcile(ctx *deploy.DeployContext) (reconcile.Result, bool, error) {
+	if m.migrationDone {
+		return reconcile.Result{}, true, nil
+	}
+
+	done, err := m.migrate(ctx)
+	if done && err == nil {
+		m.migrationDone = true
+		// Give some time for the migration resources to be flushed and rerun reconcile
+		return reconcile.Result{RequeueAfter: 5 * time.Second}, false, err
+	}
+	return reconcile.Result{}, done, err
+}
+
+func (m *Migrator) Finalize(ctx *deploy.DeployContext) error {
+	return nil
+}
+
+func (m *Migrator) migrate(ctx *deploy.DeployContext) (bool, error) {
+	if err := addRequiredLabelsForConfigMaps(ctx); err != nil {
+		return false, err
+	}
+
+	if err := addRequiredLabelsForSecrets(ctx); err != nil {
+		return false, err
+	}
+
+	if err := addRequiredLabelsForPartOfCheObjects(ctx); err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// addRequiredLabelsForConfigMaps processes the following cofig maps:
+// - 'che-git-self-signed-cert' if spec.server.gitSelfSignedCert is set to true
+// - spec.server.serverTrustStoreConfigMapName (legacy way to add trusted CA certificates)
+func addRequiredLabelsForConfigMaps(ctx *deploy.DeployContext) error {
+	// Legacy config map with additional CA certificates
+	if ctx.CheCluster.Spec.Server.ServerTrustStoreConfigMapName != "" {
+		if err := addRequiredLabelsForConfigMap(ctx, ctx.CheCluster.Spec.Server.ServerTrustStoreConfigMapName); err != nil {
+			return err
+		}
+	}
+
+	// Config map with CA certificates for git
+	if ctx.CheCluster.Spec.Server.GitSelfSignedCert {
+		if err := addRequiredLabelsForConfigMap(ctx, deploy.GitSelfSignedCertsConfigMapName); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// addRequiredLabelsForSecrets processes the following secrets:
+// - spec.k8s.tlsSecretName
+// - 'self-signed-certificate' secret
+// - spec.server.cheHostTLSSecret
+// - spec.server.proxySecret
+// - spec.database.chePostgresSecret
+// - spec.auth.identityProviderSecret
+// - spec.auth.identityProviderPostgresSecret
+// - spec.auth.oAuthSecret
```

Comment:
pls remove handling this field.
It is turned out to be a string not a secret name

Reply:
ok